### PR TITLE
batch queries for workspace computations

### PR DESCRIPTION
**Pull request by tpolecat**
This changes the `ObservationMapping` `effectField`s for `configurationRequests` and `workflow` to query returned observations at the same time rather than one at a time. It also combines some queries that were easily done together. It's kind of a big refactoring and it will be hard to review, sorry.

The goal here is to make the number of queries O(1) but there are still two places (`generatorParamsService.selectMany` and `itcService.selectAll`) where we query per-program and one (`generator.executionState`) where we query per-observation. So this is an improvement but there remains work to do.

### CODE REVIEW

File: modules/service/src/main/scala/lucuma/odb/graphql/mapping/ObservationMapping.scala [hunk 1]
```diff
@@ -132,30 +134,61 @@ trait ObservationMapping[F[_]]
...
-  lazy val configurationRequestsQueryHandler: EffectHandler[F] = {
-    val readEnv: Env => Result[Unit] = _ => ().success
-
-    val calculate: (Program.Id, Observation.Id, Unit) => F[Result[List[ConfigurationRequest]]] =
-      (_, oid, _) =>
-        services.useTransactionally {
-          configurationService
-            .selectRequests(oid)
-        }
+  lazy val configurationRequestsQueryHandler: EffectHandler[F] = { pairs =>
```
Comment (toddburnside) [inline modules/service/src/main/scala/lucuma/odb/graphql/mapping/ObservationMapping.scala:137]: I don't think I'd seen this method of creating an EffectHandler before.

File: modules/service/src/main/scala/lucuma/odb/service/ObservationService.scala [hunk 2]
```diff
@@ -171,8 +171,8 @@ object ObservationService {
       .getOrElse(GenericConstraintViolationMessage(ex.message))
 
   // observation validation messages
-  val AsterismOutOfRangeMsg = "Asterism out of Call for Proposals limits."
-  val ExplicitBaseOutOfRangeMsg = "Explicit base out of Call for Proposals limits."
+  // val AsterismOutOfRangeMsg = "Asterism out of Call for Proposals limits."
```
Comment (toddburnside) [inline modules/service/src/main/scala/lucuma/odb/service/ObservationService.scala:174]: Commented code

File: modules/service/src/main/scala/lucuma/odb/graphql/mapping/ObservationMapping.scala [hunk 3]
```diff
@@ -132,30 +134,61 @@ trait ObservationMapping[F[_]]
...
+    val sequence: ResultT[F, List[((Program.Id, Observation.Id), Cursor, Context)]] =
+      ResultT.fromResult:
+        pairs.traverse: (query, cursor) =>
+          for {
+            p <- cursor.fieldAs[Program.Id]("programId")
+            o <- cursor.fieldAs[Observation.Id]("id")
+            c <- Query.childContext(cursor.context, query)
+          } yield ((p, o), cursor, c)
+
+    // Pass the pid+oid pars to configurationService.selectRequests to get the
```
Comment (toddburnside) [inline modules/service/src/main/scala/lucuma/odb/graphql/mapping/ObservationMapping.scala:150]: Unimportant, but typo in `pars`

**Review by toddburnside — APPROVED** [review]
Wow, this was a lot of work. Thanks for taking it on! LGTM

File: modules/service/src/main/scala/lucuma/odb/graphql/mapping/ObservationMapping.scala [hunk 4]
```diff
@@ -132,30 +134,61 @@ trait ObservationMapping[F[_]]
...
-        services.use { s =>
-          s.observationWorkflowService
-            .getWorkflow(oid, commitHash, itcClient, timeEstimateCalculator)
-        }
 
-    effectHandler(readEnv, calculate)
+  lazy val workflowQueryHandler: EffectHandler[F] = { pairs =>
+
+    // Here's the ordered sequence of stuff we need to deal with: 
+    // a pid+oid pair, the parent cursor, and the child context.
```
Comment (swalker2m) [inline modules/service/src/main/scala/lucuma/odb/graphql/mapping/ObservationMapping.scala:169]: This one doesn't use a `pid`, it seems.

File: modules/service/src/main/scala/lucuma/odb/graphql/mapping/ObservationMapping.scala (hunk 1 above)
Comment (tpolecat) [reply modules/service/src/main/scala/lucuma/odb/graphql/mapping/ObservationMapping.scala:137]: `EffectHandler` is a single abstract method (SAM) interface so you can define it using function syntax. Scala has supported this for a long time but it's not commonly used.

Comment (swalker2m) [reply modules/service/src/main/scala/lucuma/odb/graphql/mapping/ObservationMapping.scala:137]: I only recently figured out that's also what's going on with `Matcher` implementations...

**Comment by tpolecat** [comment]
(Added a test to make sure we actually get back different workflows for different observations.)

**Pull request merged**